src/main/java/com/metamx/common/CompressionUtils.java

```diff
@@ -0,0 +1,388 @@
+        }
+        zipOut.putNextEntry(new ZipEntry(file.getName()));
+        totalSize += Files.asByteSource(file).copyTo(zipOut);
+      }
+      zipOut.closeEntry();
+    }
+
+    return totalSize;
+  }
+
+  public static FileUtils.FileCopyResult unzip(
+      final ByteSource byteSource,
+      final File outDir,
+      final Predicate<Throwable> shouldRetry,
+      boolean cacheLocally
+  ) throws IOException
+  {
+    if (!cacheLocally) {
+      try {
+        return RetryUtils.retry(
+            new Callable<FileUtils.FileCopyResult>()
+            {
+              @Override
+              public FileUtils.FileCopyResult call() throws Exception
+              {
+                return unzip(byteSource.openStream(), outDir);
+              }
+            },
+            shouldRetry,
+            DEFAULT_RETRY_COUNT
+        );
+      }
+      catch (Exception e) {
+        throw Throwables.propagate(e);
+      }
+    }
+    final File tmpFile = File.createTempFile("compressionUtilZipCache", ZIP_SUFFIX);
+    try {
+      FileUtils.FileCopyResult copyResult = FileUtils.retryCopy(
+          byteSource,
+          tmpFile,
+          shouldRetry,
+          DEFAULT_RETRY_COUNT
+      );
+      return unzip(tmpFile, outDir);
+    }
+    finally {
+      if (!tmpFile.delete()) {
+        log.warn("Could not delete zip cache at [%s]", tmpFile.toString());
+      }
+    }
+  }
+
+  public static FileUtils.FileCopyResult unzip(
+      final ByteSource byteSource,
+      final File outDir,
+      boolean cacheLocally
+  ) throws IOException
+  {
+    return unzip(byteSource, outDir, Predicates.<Throwable>alwaysTrue(), cacheLocally);
+  }
+
+  public static FileUtils.FileCopyResult unzip(final File pulledFile, final File outDir) throws IOException
+  {
+    if (!(outDir.exists() && outDir.isDirectory())) {
+      throw new ISE("outDir[%s] must exist and be a directory", outDir);
+    }
+    log.info("Unzipping file[%s] to [%s]", pulledFile, outDir);
+    final FileUtils.FileCopyResult result = new FileUtils.FileCopyResult();
+    try (final ZipFile zipFile = new ZipFile(pulledFile)) {
+      final Enumeration<? extends ZipEntry> enumeration = zipFile.entries();
+      while (enumeration.hasMoreElements()) {
+        final ZipEntry entry = enumeration.nextElement();
+        result.addFiles(
+            FileUtils.retryCopy(
+                new ByteSource()
+                {
+                  @Override
+                  public InputStream openStream() throws IOException
+                  {
+                    return new BufferedInputStream(zipFile.getInputStream(entry));
+                  }
+                },
+                new File(outDir, entry.getName()),
+                Predicates.<Throwable>alwaysTrue(),
+                DEFAULT_RETRY_COUNT
+            ).getFiles()
+        );
+      }
+    }
+    return result;
+  }
+
+  // Use unzip(ByteStream, File) if possible
```

Comment:
No because there are many cases where a ByteStream would not be available. @xvrl and I had a discussion about that previously